distributed/deploy/adaptive.py

```diff
@@ -28,49 +33,154 @@ class Adaptive(object):
     ...         """ Bring worker count up to n """
     ...     def scale_down(self, workers):
     ...        """ Remove worker addresses from cluster """
+
+    Notes
+    -----
+    Subclasses can override :meth:`Adaptive.should_scale_up` and
+    :meth:`Adaptive.should_scale_down` to control when the cluster should be
+    resized. The default implementation checks if there are too many tasks
+    per worker or too little memory available (see :meth:`Adaptive.needs_cpu`
+    and :meth:`Adaptive.needs_memory`).
+
+    :meth:`Adaptive.get_scale_up_kwargs` method controls the arguments passed to
+    the cluster's ``scale_up`` method.
     '''
-    def __init__(self, scheduler, cluster, interval=1000, startup_cost=1):
+
+    def __init__(self, scheduler, cluster, interval=1000, startup_cost=1,
+                 scale_factor=2):
         self.scheduler = scheduler
         self.cluster = cluster
         self.startup_cost = startup_cost
+        self.scale_factor = scale_factor
         self._adapt_callback = PeriodicCallback(self._adapt, interval,
                                                 self.scheduler.loop)
         self.scheduler.loop.add_callback(self._adapt_callback.start)
         self._adapting = False
 
+    def needs_cpu(self):
+        """
+        Check if the cluster is CPU constrained (too many tasks per core)
+
+        Notes
+        -----
+        Returns ``True`` if the occupancy per core is some factor larger
+        than ``startup_cost``.
+        """
+        total_occupancy = self.scheduler.total_occupancy
+        total_cores = sum(self.scheduler.ncores.values())
+
+        if total_occupancy / (total_cores + 1e-9) > self.startup_cost * 2:
+            logger.info("CPU limit exceeded [%d occupancy / %d cores]",
+                        total_occupancy, total_cores)
+            return True
+        else:
+            return False
+
+    def needs_memory(self):
+        """
+        Check if the cluster is RAM constrained
+
+        Notes
+        -----
+        Returns ``True`` if  the required bytes in distributed memory is some
+        factor larger than the actual distributed memory available.
+        """
+        limit_bytes = {w: self.scheduler.worker_info[w]['memory_limit']
+                        for w in self.scheduler.worker_info}
+        worker_bytes = self.scheduler.worker_bytes
+
+        limit = sum(limit_bytes.values())
+        total = sum(worker_bytes.values())
+        if total > 0.6 * limit:
+            logger.info("Ram limit exceeded [%d/%d]", limit, total)
+            return True
+        else:
+            return False
```

Comment:
> More generally this brings up the thought that maybe we should specify targets rather than ask for more/less. This is beyond the scope of this PR though.

Agreed that targets are probably better. I'm essentially doing that via the `get_scale_up_kwargs`: in the dask-drmaa PR on top of this: https://github.com/TomAugspurger/dask-drmaa/blob/8021d288fee20d5e464532505753c0fe74583ba8/dask_drmaa/adaptive.py#L68